<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="ClusterDE">
<title>Perform ClusterDE on a cell line dataset • ClusterDE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform ClusterDE on a cell line dataset">
<meta property="og:description" content="ClusterDE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ClusterDE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ClusterDE-PBMC.html">Perform ClusterDE on a PBMC dataset</a>
    <a class="dropdown-item" href="../articles/ClusterDE-cellline.html">Perform ClusterDE on a cell line dataset</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SONGDONGYUAN1994/ClusterDE/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Perform ClusterDE on a cell line dataset</h1>
                        <h4 data-toc-skip class="author">Dongyuan
Song</h4>
            <address class="author_afil">
      Bioinformatics IDP, University of California, Los
Angeles<br><a class="author_email" href="mailto:#"></a><a href="mailto:dongyuansong@ucla.edu" class="email">dongyuansong@ucla.edu</a>
      </address>
                  
            <h4 data-toc-skip class="date">16 October 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/SONGDONGYUAN1994/ClusterDE/blob/HEAD/vignettes/ClusterDE-cellline.Rmd" class="external-link"><code>vignettes/ClusterDE-cellline.Rmd</code></a></small>
      <div class="d-none name"><code>ClusterDE-cellline.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#tools::R_user_dir("ClusterDE", which="cache")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SONGDONGYUAN1994/ClusterDE" class="external-link">ClusterDE</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="download-data">Download data<a class="anchor" aria-label="anchor" href="#download-data"></a>
</h2>
<p>We download the a cell line data set H2228. The original data is from
<a href="https://github.com/LuyiTian/sc_mixology" class="external-link">Tian et al., Nature
Methods 2019</a>. Since the data is from the pure cell line, it should
not have cell types, and, of course, between cell type DE genes.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://figshare.com/ndownloader/files/41395260"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/as.Seurat.html" class="external-link">as.Seurat</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Keys should be one or more alphanumeric characters followed by an</span></span>
<span><span class="co">#&gt; underscore, setting key from PC to PC_</span></span>
<span><span class="co">#&gt; Warning: All keys should be one or more alphanumeric characters followed by an</span></span>
<span><span class="co">#&gt; underscore '_', setting key to PC_</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-the-regular-seurat-pipeline">Run the regular Seurat pipeline<a class="anchor" aria-label="anchor" href="#run-the-regular-seurat-pipeline"></a>
</h2>
<p>We perform the default Seurat clustering. Please note that ClusterDE
is designed for 1 vs 1 comparison; therefore, we set the resolution as
0.2 here to obtain two clusters for illustration purpose.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">RNGkind</a></span><span class="op">(</span><span class="st">"L'Ecuyer-CMRG"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline</span><span class="op">)</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline</span><span class="op">)</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline</span><span class="op">)</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline</span><span class="op">)</span></span>
<span><span class="co">#&gt; PC_ 1 </span></span>
<span><span class="co">#&gt; Positive:  RPS14, RPL18AP3, RPL36, RPS23, RPL28, LRRC75A-AS1, AC079250.1, FTH1, ZFAS1, EEF2 </span></span>
<span><span class="co">#&gt;     RPL7P9, RPL13A, RPS3AP26, EEF1A1P13, RPS16, RPS23P8, RPL13AP5, RPL29, FTH1P10, RPL13AP25 </span></span>
<span><span class="co">#&gt;     SNHG5, FTH1P8, RPL4, RPS3AP6, AC064799.1, C6orf48, FTH1P7, C1orf56, RPL7AP6, TMSB4X </span></span>
<span><span class="co">#&gt; Negative:  PSMB2, PSMA7, U2AF1, NUDC, RBM8A, CALM1, BUB3, CLIC1, U2AF1L5, XRCC5 </span></span>
<span><span class="co">#&gt;     VPS29, RBM8B, CACYBP, RPA3, SSBP1, PSMC5, MRPL47, PSMD8, BRIX1, CNIH4 </span></span>
<span><span class="co">#&gt;     PCMT1, PSMD13, CYC1, PRDX2, SEPT7, S100A11, VDAC3, PSME2P2, ZWINT, HMGB1 </span></span>
<span><span class="co">#&gt; PC_ 2 </span></span>
<span><span class="co">#&gt; Positive:  NACA, RPL7AP6, SKP1, UBA52, BTF3, SSR2, RPL7A, ARPC3, RPL9P9, PPIA </span></span>
<span><span class="co">#&gt;     PSMD4, EIF1, RPL10, LGALS3BP, RPL10P16, SNRPB2, RPL10P9, S100A11, PPIB, ANXA5 </span></span>
<span><span class="co">#&gt;     EEF2, PSME1, SSBP1, SSR4, RPL7P9, COPE, BSG, MGST1, VPS28, COPS6 </span></span>
<span><span class="co">#&gt; Negative:  SIVA1, HNRNPAB, RPL39L, DEK, CDCA5, TMPO, FAM111A, ASF1B, CENPK, ESCO2 </span></span>
<span><span class="co">#&gt;     BRCA1, H2AFV, RAD51AP1, MT-RNR2, ORC6, CENPX, SNRNP25, FBXO5, RRM1, DIAPH3 </span></span>
<span><span class="co">#&gt;     USP1, CDCA4, TMEM106C, PGP, LSM4, C21orf58, CENPN, BRI3BP, SGO1, CHAF1A </span></span>
<span><span class="co">#&gt; PC_ 3 </span></span>
<span><span class="co">#&gt; Positive:  RPL13AP5, RPL13AP25, AC024293.1, RPL29, RPSAP19, RPS5, RPL18, RPS3AP26, RPS3AP6, RPL15 </span></span>
<span><span class="co">#&gt;     RPL28, DRAP1, RPS11, RPL9P9, RPL13AP7, RPS19, DCBLD2, FXYD5, FEN1, SLBP </span></span>
<span><span class="co">#&gt;     RPS15, COTL1, RPSA, FLNA, RPL7AP6, RPL36, C1orf21, CPA4, ORC6, RPS16 </span></span>
<span><span class="co">#&gt; Negative:  SMIM22, TSPAN13, ST14, PERP, CRB3, MT-CO1, SERINC2, ATP1B1, CDH1, F11R </span></span>
<span><span class="co">#&gt;     B2M, MT-RNR2, SPINT1, NMB, PLA2G16, SPDEF, CD55, ADGRF1, TSPAN1, LIMA1 </span></span>
<span><span class="co">#&gt;     ERBB3, ERO1A, ASS1, CDA, ALCAM, SYNGR2, MT-CO2, CDH3, C3, LSR </span></span>
<span><span class="co">#&gt; PC_ 4 </span></span>
<span><span class="co">#&gt; Positive:  IFNGR1, NAMPT, NAP1L1, CPD, LMAN1, CALR, ITGA2, NAMPTP1, ITM2B, RRM1 </span></span>
<span><span class="co">#&gt;     C3, RHOBTB3, CTHRC1, EEF2, HSD17B11, C1S, IFI16, SMC2, CPE, EPHX1 </span></span>
<span><span class="co">#&gt;     DST, HLA-DMB, NUCB2, MT-ND6, TMEM45A, BRCA1, CDK5RAP2, HINT1, C1R, FAM111A </span></span>
<span><span class="co">#&gt; Negative:  S100A16, TMA7, TIMM8B, PFN1, SLIRP, GPX1, LAMC2, POLR2L, MRPL52, RPS19 </span></span>
<span><span class="co">#&gt;     CDH1, TOMM40, ATP5MD, HSPE1, NAA10, GPX1P1, RPS16, MRPL12, MCRIP2, PDCD5 </span></span>
<span><span class="co">#&gt;     RPL18, PLEC, S100A13, RPL36AL, LAD1, MGLL, BOLA2B, MISP, MRPL36, SEC61G </span></span>
<span><span class="co">#&gt; PC_ 5 </span></span>
<span><span class="co">#&gt; Positive:  HINT1, COX5B, TXN, SOD1, NDUFA4, NDUFS6, ATP5PO, ATP5MC3, S100A10, ATP6V0E1 </span></span>
<span><span class="co">#&gt;     CYB5A, SLIRP, NDUFB4, ATP5MC1, HSPE1, TXNP6, POMP, POLR2L, RPS14, HSPE1P4 </span></span>
<span><span class="co">#&gt;     RPS15, CBR1, NDUFB3, HSPE1P3, ATP5PD, COX7B, ADGRF1, PSMB9, AC079250.1, NDUFAB1 </span></span>
<span><span class="co">#&gt; Negative:  BTG1, PPP1R15A, EIF1, JUN, CEBPG, H3F3B, TMEM132A, C6orf48, HIST2H4B, SGK1 </span></span>
<span><span class="co">#&gt;     KPNA4, PMEPA1, KLF6, CDKN1A, WARS, PEA15, GARS, MAP1LC3B, SNHG12, SERTAD1 </span></span>
<span><span class="co">#&gt;     LAMC2, EPB41L4A-AS1, NAP1L1, SNHG5, KLF10, SLC7A5, KIF5B, ATP2B1, EIF5, ABL2</span></span>
<span><span class="co">#&gt; Warning: Cannot add objects with duplicate keys (offending key: PC_), setting</span></span>
<span><span class="co">#&gt; key to 'pca_'</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; Computing SNN</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline</span>, resolution <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 758</span></span>
<span><span class="co">#&gt; Number of edges: 24895</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8257</span></span>
<span><span class="co">#&gt; Number of communities: 2</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="va">cellline</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="co">#&gt; 01:08:39 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; 01:08:39 Read 758 rows and found 10 numeric columns</span></span>
<span><span class="co">#&gt; 01:08:39 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; 01:08:39 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 01:08:39 Writing NN index file to temp file /tmp/RtmpKj5gZZ/file1cc72b97d744</span></span>
<span><span class="co">#&gt; 01:08:39 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 01:08:39 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 01:08:40 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 01:08:41 Initializing from normalized Laplacian + noise (using irlba)</span></span>
<span><span class="co">#&gt; 01:08:41 Commencing optimization for 500 epochs, with 27914 positive edges</span></span>
<span><span class="co">#&gt; 01:08:42 Optimization finished</span></span>
<span><span class="co">#&gt; Warning: Cannot add objects with duplicate keys (offending key: UMAP_), setting</span></span>
<span><span class="co">#&gt; key to 'umap_'</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Clustering result"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClusterDE-cellline_files/figure-html/unnamed-chunk-3-1.png" width="700">
From the UMAP, the two clusters seem to be dubious. Although we do not
expect the existence of cell types, when we perform Seurat DE test
between the two clusters, and we get &gt; 1000 genes with FDR &lt; 0.05.
It means that the double dipping introduces a huge number of
discoveries.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">original_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span><span class="va">cellline</span>, </span>
<span>                                ident.1 <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                                ident.2 <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                                min.pct <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                                logfc.threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">#&gt; For a more efficient implementation of the Wilcoxon Rank Sum Test,</span></span>
<span><span class="co">#&gt; (default method for FindMarkers) please install the limma package</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; install.packages('BiocManager')</span></span>
<span><span class="co">#&gt; BiocManager::install('limma')</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; After installation of limma, Seurat will automatically use the more </span></span>
<span><span class="co">#&gt; efficient implementation (no further action necessary).</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">original_markers</span><span class="op">$</span><span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1023</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="generate-synthetic-null-data">Generate synthetic null data<a class="anchor" aria-label="anchor" href="#generate-synthetic-null-data"></a>
</h2>
<p>We first generate the synthetic null data based on the target data
(real data). You can increase the number of cores to speed it up.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">synthetic_null</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="../reference/constructNull.html">constructNull</a></span><span class="op">(</span><span class="va">count_mat</span>, nCores <span class="op">=</span> <span class="fl">2</span>, fastVersion <span class="op">=</span> <span class="cn">TRUE</span>, family <span class="op">=</span> <span class="st">"nb"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt; 173.531  44.039 115.019</span></span></code></pre></div>
<p>We perform the same pipeline as we did for target data. Please note
we need two clusters here, too. This is straightforward for clustering
method with predefined cluster numbers, e.g., K-means. For louvain
clustering used in Seurat you may manually adjust the resolution
parameter so that we still get two clusters.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellline_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">synthetic_null</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">cellline_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline_null</span><span class="op">)</span></span>
<span><span class="va">cellline_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline_null</span><span class="op">)</span></span>
<span><span class="va">cellline_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline_null</span><span class="op">)</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="va">cellline_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline_null</span><span class="op">)</span></span>
<span><span class="co">#&gt; PC_ 1 </span></span>
<span><span class="co">#&gt; Positive:  PSMB2, PSMC5, PSMA7, XRCC5, SKP1, PSMD8, NUDC, PSME2P2, U2AF1, VPS29 </span></span>
<span><span class="co">#&gt;     CALM1, MRPL47, MGST1, CLIC1, CNIH4, SSBP1, RBM8A, TAGLN2, RPA3, COX5A </span></span>
<span><span class="co">#&gt;     PSME2, U2AF1L5, SNRPF, ARPC3, S100A11, PPIA, ATP5PD, MYL12B, RBM8B, ATP5PO </span></span>
<span><span class="co">#&gt; Negative:  LRRC75A-AS1, RPS14, FTH1, RPL29, RPL28, ZFAS1, RPL13A, FTH1P10, FTH1P7, RPS16 </span></span>
<span><span class="co">#&gt;     EEF2, SNHG5, RPL13AP5, C1orf56, TMSB10, FTH1P8, MKNK2, STEAP2, TMSB4X, RPL32 </span></span>
<span><span class="co">#&gt;     ASMTL, RPL13AP25, RPL4, NBPF10, C6orf48, NREP, VCAN, RPL9P9, ANKRD36B, HCFC1R1 </span></span>
<span><span class="co">#&gt; PC_ 2 </span></span>
<span><span class="co">#&gt; Positive:  SIVA1, RPL39L, CENPK, TMPO, HNRNPAB, DEK, ASF1B, H2AFV, CENPX, SNRNP25 </span></span>
<span><span class="co">#&gt;     CKLF, TMEM106C, CHCHD2, CDCA5, FAM111A, RAD51AP1, RRM1, ZWINT, LSM4, SAC3D1 </span></span>
<span><span class="co">#&gt;     CDCA4, BRI3BP, CENPN, SUZ12, BRCA1, PGP, ESCO2, ORC6, FBXO5, PRADC1 </span></span>
<span><span class="co">#&gt; Negative:  NACA, RPL7AP6, RPL10, RPL7A, RPL10P16, RPL9P9, UBA52, BTF3, RPL10P9, SSR2 </span></span>
<span><span class="co">#&gt;     EEF2, SKP1, RPL4, RPL18, LRRC75A-AS1, ANXA5, RPS4XP11, BSG, PSMD4, PPIB </span></span>
<span><span class="co">#&gt;     APEX1, ARPC3, EIF3M, SNRPB2, MYDGF, VPS28, CTSD, EDF1, PNRC1, EIF1 </span></span>
<span><span class="co">#&gt; PC_ 3 </span></span>
<span><span class="co">#&gt; Positive:  RPL29, RPL14, RPSAP19, RPL15, RPL13AP25, RPL32, RPL18, RPL13AP5, RPSA, DRAP1 </span></span>
<span><span class="co">#&gt;     RPS5, RPS11, TMA7, CFL1, RPS19, FLNA, DCBLD2, RPL28, PFN1, RPS16 </span></span>
<span><span class="co">#&gt;     PDCD5, SLBP, RPS15, PEBP1, AP2S1, FXYD5, BASP1, TMSB10, NME1, HSPE1P4 </span></span>
<span><span class="co">#&gt; Negative:  ASS1, ADGRF1, C3, B2M, PERP, ATP1B1, CASP4, TSPAN13, CPD, LIPH </span></span>
<span><span class="co">#&gt;     SMIM22, EPS8, IFI16, GOLGB1, TMED10, NMB, SDC4, LIMA1, NMI, SPINT1 </span></span>
<span><span class="co">#&gt;     TNFAIP2, CCDC186, CD55, HLA-DMB, F11R, NAMPT, SELENBP1, SQSTM1, SQOR, PLA2G16 </span></span>
<span><span class="co">#&gt; PC_ 4 </span></span>
<span><span class="co">#&gt; Positive:  S100A16, H3F3B, LAMC2, RPL36AL, PMEPA1, EIF5, TBRG1, RPL15, RNF145, ABRACL </span></span>
<span><span class="co">#&gt;     IMP4, SERTAD1, LAMB3, KDM5B, HMGB3, HIST1H2BD, DDX21, ZNF593, ZPR1, PFDN2 </span></span>
<span><span class="co">#&gt;     HIST2H4B, TSSC4, GOT1, PLEC, HIST3H2A, ITGA6, OTUB1, ALCAM, COA6, EIF1 </span></span>
<span><span class="co">#&gt; Negative:  IFNGR1, RFC4, NAMPT, BRCA1, EID1, POLD3, NDUFA4, HINT1, MCM3, ATP6V0E1 </span></span>
<span><span class="co">#&gt;     RPA3, SOD1, CYB5A, DNMT1, CDT1, GMNN, CHAF1A, THYN1, CPD, CXXC5 </span></span>
<span><span class="co">#&gt;     ADGRF1, ORC6, NAMPTP1, MCM6, CARHSP1, ACOT13, RPL7AP6, RHOBTB3, RMI2, NCOA7 </span></span>
<span><span class="co">#&gt; PC_ 5 </span></span>
<span><span class="co">#&gt; Positive:  MT-ND6, RAD21, DIAPH3, NDC80, ITGA2, C6orf48, KIF5B, NUCKS1, TMEM132A, RRM1 </span></span>
<span><span class="co">#&gt;     ANP32E, NCAPG, CCDC88A, SPC25, ZNF622, SMC2, TATDN1P1, GARS, MTHFD2, CEP55 </span></span>
<span><span class="co">#&gt;     C1orf56, SF3B2, VCAN, CEP57, CKAP2L, PIGK, NCAPD3, BCLAF1P2, TBC1D7, RTCB </span></span>
<span><span class="co">#&gt; Negative:  UQCRQ, NDUFS6, COX6B1, FTH1, FTH1P8, ATP5MC3, FTH1P10, APRT, COX7B, COX4I1 </span></span>
<span><span class="co">#&gt;     ATP6V0E1, FTH1P7, NDUFB3, POLR2L, LY6E, SLIRP, PFN1, NMB, CHCHD10, HINT1 </span></span>
<span><span class="co">#&gt;     ATP5MF, POMP, NDUFB4, MCRIP2, RPL36AL, MRPL23, S100A10, SEC61G, GPX4, DUSP23</span></span>
<span><span class="va">cellline_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline_null</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; Computing SNN</span></span>
<span><span class="va">cellline_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline_null</span>, resolution <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 758</span></span>
<span><span class="co">#&gt; Number of edges: 29614</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.7309</span></span>
<span><span class="co">#&gt; Number of communities: 2</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="va">cellline_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline_null</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; 01:12:05 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; 01:12:05 Read 758 rows and found 10 numeric columns</span></span>
<span><span class="co">#&gt; 01:12:05 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; 01:12:05 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 01:12:05 Writing NN index file to temp file /tmp/RtmpKj5gZZ/file1cc778fbf7ec</span></span>
<span><span class="co">#&gt; 01:12:05 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 01:12:05 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 01:12:06 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 01:12:07 Initializing from normalized Laplacian + noise (using irlba)</span></span>
<span><span class="co">#&gt; 01:12:07 Commencing optimization for 500 epochs, with 28370 positive edges</span></span>
<span><span class="co">#&gt; 01:12:08 Optimization finished</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellline_null</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClusterDE-cellline_files/figure-html/unnamed-chunk-6-1.png" width="700">
We perform the DE test on synthetic null data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">null_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span><span class="va">cellline_null</span>, </span>
<span>                            ident.1 <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                            ident.2 <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                            min.pct <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                            logfc.threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>We extract the p-values from both original data and synthetic null
data, then use <code>ClusterDE</code> to “compare” them. We do not
discover any DE genes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">original_pval</span> <span class="op">&lt;-</span> <span class="va">original_markers</span><span class="op">$</span><span class="va">p_val</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">original_pval</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">original_markers</span><span class="op">)</span></span>
<span></span>
<span><span class="va">null_pval</span> <span class="op">&lt;-</span> <span class="va">null_markers</span><span class="op">$</span><span class="va">p_val</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">null_pval</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">null_markers</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="../reference/callDE.html">callDE</a></span><span class="op">(</span><span class="va">original_pval</span>, <span class="va">null_pval</span>, nlogTrans <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">$</span><span class="va">DEgenes</span></span>
<span><span class="co">#&gt; character(0)</span></span></code></pre></div>
<p>We can also visualize the distribution of contrast scores (diff
between the -log p-values from real and null). It is roughly symmetric
around 0.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">summaryTable</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cs</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"white"</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Distribution of constrast scores"</span><span class="op">)</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span>
<span><span class="co">#&gt; Warning: Removed 1 rows containing non-finite values (`stat_bin()`).</span></span></code></pre></div>
<p><img src="ClusterDE-cellline_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Random number generation:</span></span>
<span><span class="co">#&gt;  RNG:     L'Ecuyer-CMRG </span></span>
<span><span class="co">#&gt;  Normal:  Inversion </span></span>
<span><span class="co">#&gt;  Sample:  Rejection </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] ggplot2_3.4.4               SingleCellExperiment_1.22.0</span></span>
<span><span class="co">#&gt;  [3] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">#&gt;  [5] GenomicRanges_1.52.1        GenomeInfoDb_1.36.4        </span></span>
<span><span class="co">#&gt;  [7] IRanges_2.34.1              S4Vectors_0.38.2           </span></span>
<span><span class="co">#&gt;  [9] BiocGenerics_0.46.0         MatrixGenerics_1.12.3      </span></span>
<span><span class="co">#&gt; [11] matrixStats_1.0.0           SeuratObject_4.1.4         </span></span>
<span><span class="co">#&gt; [13] Seurat_4.4.0                ClusterDE_0.99.1           </span></span>
<span><span class="co">#&gt; [15] BiocStyle_2.28.1           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3      jsonlite_1.8.7          magrittr_2.0.3         </span></span>
<span><span class="co">#&gt;   [4] spatstat.utils_3.0-3    farver_2.1.1            rmarkdown_2.25         </span></span>
<span><span class="co">#&gt;   [7] zlibbioc_1.46.0         fs_1.6.3                ragg_1.2.6             </span></span>
<span><span class="co">#&gt;  [10] vctrs_0.6.4             ROCR_1.0-11             memoise_2.0.1          </span></span>
<span><span class="co">#&gt;  [13] spatstat.explore_3.2-3  RCurl_1.98-1.12         S4Arrays_1.0.6         </span></span>
<span><span class="co">#&gt;  [16] htmltools_0.5.6.1       sass_0.4.7              sctransform_0.4.0      </span></span>
<span><span class="co">#&gt;  [19] parallelly_1.36.0       KernSmooth_2.23-21      bslib_0.5.1            </span></span>
<span><span class="co">#&gt;  [22] htmlwidgets_1.6.2       desc_1.4.2              ica_1.0-3              </span></span>
<span><span class="co">#&gt;  [25] plyr_1.8.9              plotly_4.10.2           zoo_1.8-12             </span></span>
<span><span class="co">#&gt;  [28] cachem_1.0.8            igraph_1.5.1            mime_0.12              </span></span>
<span><span class="co">#&gt;  [31] lifecycle_1.0.3         pkgconfig_2.0.3         Matrix_1.6-1.1         </span></span>
<span><span class="co">#&gt;  [34] R6_2.5.1                fastmap_1.1.1           GenomeInfoDbData_1.2.10</span></span>
<span><span class="co">#&gt;  [37] fitdistrplus_1.1-11     future_1.33.0           shiny_1.7.5            </span></span>
<span><span class="co">#&gt;  [40] digest_0.6.33           colorspace_2.1-0        patchwork_1.1.3        </span></span>
<span><span class="co">#&gt;  [43] rprojroot_2.0.3         tensor_1.5              irlba_2.3.5.1          </span></span>
<span><span class="co">#&gt;  [46] kde1d_1.0.5             textshaping_0.3.7       labeling_0.4.3         </span></span>
<span><span class="co">#&gt;  [49] progressr_0.14.0        fansi_1.0.5             spatstat.sparse_3.0-2  </span></span>
<span><span class="co">#&gt;  [52] httr_1.4.7              polyclip_1.10-6         abind_1.4-5            </span></span>
<span><span class="co">#&gt;  [55] compiler_4.3.1          withr_2.5.1             MASS_7.3-60            </span></span>
<span><span class="co">#&gt;  [58] DelayedArray_0.26.7     tools_4.3.1             lmtest_0.9-40          </span></span>
<span><span class="co">#&gt;  [61] httpuv_1.6.11           future.apply_1.11.0     goftest_1.2-3          </span></span>
<span><span class="co">#&gt;  [64] glue_1.6.2              nlme_3.1-162            promises_1.2.1         </span></span>
<span><span class="co">#&gt;  [67] grid_4.3.1              Rtsne_0.16              cluster_2.1.4          </span></span>
<span><span class="co">#&gt;  [70] reshape2_1.4.4          generics_0.1.3          gtable_0.3.4           </span></span>
<span><span class="co">#&gt;  [73] spatstat.data_3.0-1     tidyr_1.3.0             data.table_1.14.8      </span></span>
<span><span class="co">#&gt;  [76] XVector_0.40.0          sp_2.1-0                utf8_1.2.3             </span></span>
<span><span class="co">#&gt;  [79] spatstat.geom_3.2-5     RcppAnnoy_0.0.21        ggrepel_0.9.3          </span></span>
<span><span class="co">#&gt;  [82] RANN_2.6.1              pillar_1.9.0            stringr_1.5.0          </span></span>
<span><span class="co">#&gt;  [85] later_1.3.1             splines_4.3.1           dplyr_1.1.3            </span></span>
<span><span class="co">#&gt;  [88] lattice_0.21-8          survival_3.5-5          deldir_1.0-9           </span></span>
<span><span class="co">#&gt;  [91] tidyselect_1.2.0        miniUI_0.1.1.1          pbapply_1.7-2          </span></span>
<span><span class="co">#&gt;  [94] knitr_1.44              gridExtra_2.3           bookdown_0.35          </span></span>
<span><span class="co">#&gt;  [97] scattermore_1.2         xfun_0.40               stringi_1.7.12         </span></span>
<span><span class="co">#&gt; [100] lazyeval_0.2.2          yaml_2.3.7              evaluate_0.22          </span></span>
<span><span class="co">#&gt; [103] codetools_0.2-19        tibble_3.2.1            BiocManager_1.30.22    </span></span>
<span><span class="co">#&gt; [106] cli_3.6.1               uwot_0.1.16             xtable_1.8-4           </span></span>
<span><span class="co">#&gt; [109] reticulate_1.34.0       systemfonts_1.0.5       munsell_0.5.0          </span></span>
<span><span class="co">#&gt; [112] jquerylib_0.1.4         Rcpp_1.0.11             globals_0.16.2         </span></span>
<span><span class="co">#&gt; [115] spatstat.random_3.1-6   png_0.1-8               rngWELL_0.10-9         </span></span>
<span><span class="co">#&gt; [118] parallel_4.3.1          ellipsis_0.3.2          randtoolbox_2.0.4      </span></span>
<span><span class="co">#&gt; [121] assertthat_0.2.1        pkgdown_2.0.7           mvnfast_0.2.8          </span></span>
<span><span class="co">#&gt; [124] bitops_1.0-7            listenv_0.9.0           viridisLite_0.4.2      </span></span>
<span><span class="co">#&gt; [127] scales_1.2.1            ggridges_0.5.4          crayon_1.5.2           </span></span>
<span><span class="co">#&gt; [130] leiden_0.4.3            purrr_1.0.2             rlang_1.1.1            </span></span>
<span><span class="co">#&gt; [133] rvinecopulib_0.6.3.1.1  cowplot_1.1.1</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Dongyuan Song, Kexin Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
