<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="ClusterDE">
<title>Perform ClusterDE on a PBMC dataset • ClusterDE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform ClusterDE on a PBMC dataset">
<meta property="og:description" content="ClusterDE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ClusterDE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/ClusterDE-PBMC.html">Perform ClusterDE on a PBMC dataset</a>
    <a class="dropdown-item" href="../articles/ClusterDE-cellline.html">Perform ClusterDE on a cell line dataset</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SONGDONGYUAN1994/ClusterDE/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Perform ClusterDE on a PBMC dataset</h1>
                        <h4 data-toc-skip class="author">Dongyuan
Song</h4>
            <address class="author_afil">
      Bioinformatics IDP, University of California, Los
Angeles<br><a class="author_email" href="mailto:#"></a><a href="mailto:dongyuansong@ucla.edu" class="email">dongyuansong@ucla.edu</a>
      </address>
                  
            <h4 data-toc-skip class="date">12 October 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/SONGDONGYUAN1994/ClusterDE/blob/HEAD/vignettes/ClusterDE-PBMC.Rmd" class="external-link"><code>vignettes/ClusterDE-PBMC.Rmd</code></a></small>
      <div class="d-none name"><code>ClusterDE-PBMC.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#tools::R_user_dir("ClusterDE", which="cache")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SONGDONGYUAN1994/ClusterDE" class="external-link">ClusterDE</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="co">#library(SeuratData)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="download-data">Download data<a class="anchor" aria-label="anchor" href="#download-data"></a>
</h2>
<p>The PBMC datasets are originally from <code>SeuratData</code>. We use
one of them (10x Chromium (v3) from PBMC1 replicate). We filtered out
some lowly epxressed genes to save computational time here.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># InstallData("pbmcsca")</span></span>
<span><span class="co"># data("pbmcsca")</span></span>
<span><span class="co"># pbmc &lt;- pbmcsca[, pbmcsca@meta.data$Method=="10x Chromium (v3)" &amp; pbmcsca@meta.data$Experiment == "pbmc1"]</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># pbmc &lt;- pbmc[Matrix::rowSums(pbmc@assays$RNA@counts) &gt; 100, ]</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://figshare.com/ndownloader/files/41486283"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-the-regular-seurat-pipeline">Run the regular Seurat pipeline<a class="anchor" aria-label="anchor" href="#run-the-regular-seurat-pipeline"></a>
</h2>
<p>We perform the default Seurat clustering. Note that in real data
analysis, the cell type label is usually unknown.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">RNGkind</a></span><span class="op">(</span><span class="st">"L'Ecuyer-CMRG"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="co">#&gt; PC_ 1 </span></span>
<span><span class="co">#&gt; Positive:  IL32, CCL5, TRBC2, TRAC, CD69, CST7, RORA, CTSW, SPOCK2, ITM2A </span></span>
<span><span class="co">#&gt;     GZMM, CD247, TRBC1, C12orf75, IL7R, CD8A, CD2, LDHB, GZMA, CD7 </span></span>
<span><span class="co">#&gt;     NKG7, CD6, GZMH, CD8B, BCL11B, PRF1, LYAR, LTB, FGFBP2, TCF7 </span></span>
<span><span class="co">#&gt; Negative:  LYZ, FCN1, CLEC7A, CPVL, SERPINA1, SPI1, S100A9, AIF1, NAMPT, CSTA </span></span>
<span><span class="co">#&gt;     CTSS, MAFB, MPEG1, NCF2, VCAN, FGL2, S100A8, TYMP, CST3, LST1 </span></span>
<span><span class="co">#&gt;     CYBB, CFD, FCER1G, SLC11A1, TGFBI, GRN, CD14, PSAP, SLC7A7, MS4A6A </span></span>
<span><span class="co">#&gt; PC_ 2 </span></span>
<span><span class="co">#&gt; Positive:  RPL10, EEF1A1, TMSB10, RPS2, RPS12, RPL13, RPS18, RPS23, RPLP1, TPT1 </span></span>
<span><span class="co">#&gt;     RPS8, IL32, S100A4, PFN1, RPLP0, NKG7, ARL4C, HSPA8, CST7, ZFP36L2 </span></span>
<span><span class="co">#&gt;     ANXA1, CTSW, S100A6, LDHA, CORO1A, CD247, GZMA, CALR, S100A10, GZMM </span></span>
<span><span class="co">#&gt; Negative:  NRGN, PF4, SDPR, HIST1H2AC, MAP3K7CL, PPBP, GNG11, GPX1, TUBB1, SPARC </span></span>
<span><span class="co">#&gt;     CLU, PGRMC1, FTH1, RGS18, MARCH2, TREML1, HIST1H3H, AP003068.23, NCOA4, ACRBP </span></span>
<span><span class="co">#&gt;     TAGLN2, PRKAR2B, CD9, CA2, CMTM5, CTTN, MTURN, TMSB4X, HIST1H2BJ, TSC22D1 </span></span>
<span><span class="co">#&gt; PC_ 3 </span></span>
<span><span class="co">#&gt; Positive:  CD79A, HLA-DQA1, MS4A1, LINC00926, IGHM, BANK1, IGHD, TNFRSF13C, HLA-DQB1, CD74 </span></span>
<span><span class="co">#&gt;     IGKC, HLA-DRA, BLK, CD83, CD37, CD22, ADAM28, JUND, NFKBID, HLA-DRB1 </span></span>
<span><span class="co">#&gt;     P2RX5, CD79B, VPREB3, IGLC2, FCER2, RPS8, LTB, RPS23, TCOF1, GNG7 </span></span>
<span><span class="co">#&gt; Negative:  CCL5, TMSB4X, SRGN, NKG7, ACTB, CST7, GZMH, FGFBP2, CTSW, PRF1 </span></span>
<span><span class="co">#&gt;     GZMA, GZMB, C12orf75, S100A4, ANXA1, KLRD1, NRGN, GNLY, GZMM, IL32 </span></span>
<span><span class="co">#&gt;     PF4, SDPR, PPBP, MYO1F, CD247, GAPDH, MAP3K7CL, HIST1H2AC, GNG11, TUBB1 </span></span>
<span><span class="co">#&gt; PC_ 4 </span></span>
<span><span class="co">#&gt; Positive:  FCGR3A, GZMB, FGFBP2, GZMH, NKG7, HLA-DPA1, PRF1, HLA-DPB1, CST7, GNLY </span></span>
<span><span class="co">#&gt;     KLRD1, HLA-DRB1, GZMA, CCL5, SPON2, ADGRG1, CTSW, ZEB2, PRSS23, IFITM2 </span></span>
<span><span class="co">#&gt;     CCL4, CD74, KLRF1, RHOC, MTSS1, CDKN1C, CD79B, CEP78, HLA-DQA1, CLIC3 </span></span>
<span><span class="co">#&gt; Negative:  IL7R, LEPROTL1, LTB, RCAN3, MAL, LEF1, TCF7, ZFP36L2, CAMK4, VIM </span></span>
<span><span class="co">#&gt;     LDHB, NOSIP, JUNB, SLC2A3, TRABD2A, RGCC, SATB1, TNFAIP3, TMEM123, SOCS3 </span></span>
<span><span class="co">#&gt;     AQP3, BCL11B, NELL2, TNFRSF25, CD28, PABPC1, DNAJB1, TRAT1, OXNAD1, TRAC </span></span>
<span><span class="co">#&gt; PC_ 5 </span></span>
<span><span class="co">#&gt; Positive:  CDKN1C, HES4, CSF1R, CKB, ZNF703, TCF7L2, CTSL, MS4A7, PAG1, FAM110A </span></span>
<span><span class="co">#&gt;     SIGLEC10, LRRC25, FCGR3A, LTB, RNASET2, CDH23, IL7R, RRAS, LINC01272, IFITM3 </span></span>
<span><span class="co">#&gt;     LST1, LILRB2, PILRA, RHOC, SLC2A6, PECAM1, CAMK1, TAGLN, IFI30, BID </span></span>
<span><span class="co">#&gt; Negative:  VCAN, S100A12, S100A8, CD14, CSF3R, ITGAM, CST7, GZMB, MT-CO1, GNLY </span></span>
<span><span class="co">#&gt;     KLRD1, PRF1, MS4A6A, GZMH, FGFBP2, CD93, EGR1, NKG7, S100A9, MT-CO3 </span></span>
<span><span class="co">#&gt;     IER3, THBS1, RNASE6, CLEC4E, MGST1, CTSW, SGK1, GZMA, RP11-1143G9.4, CH17-373J23.1</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; Computing SNN</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, resolution <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 3222</span></span>
<span><span class="co">#&gt; Number of edges: 108605</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.9363</span></span>
<span><span class="co">#&gt; Number of communities: 10</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="co">#&gt; 18:15:41 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; 18:15:41 Read 3222 rows and found 10 numeric columns</span></span>
<span><span class="co">#&gt; 18:15:41 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; 18:15:41 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 18:15:41 Writing NN index file to temp file /tmp/RtmptRhvxT/file1b9617566ccd</span></span>
<span><span class="co">#&gt; 18:15:41 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 18:15:42 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 18:15:43 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 18:15:44 Initializing from normalized Laplacian + noise (using irlba)</span></span>
<span><span class="co">#&gt; 18:15:45 Commencing optimization for 500 epochs, with 126446 positive edges</span></span>
<span><span class="co">#&gt; 18:15:49 Optimization finished</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Clustering result"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"CellType"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="ClusterDE-PBMC_files/figure-html/unnamed-chunk-3-1.png" width="1152"></p>
<p>In this vignette, we are interested in cluster 2 vs 8, which
approximately represent CD14+/CD16+ monocytes. Please note that
ClusterDE is designed for 1 vs 1 comparison. Therefore, users may (1)
choose the two interested clusters manually based on their knowledge or
(2) use the two closest clusters from computation (e.g.,
<code>BuildClusterTree</code> in <code>Seurat</code>).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/BuildClusterTree.html" class="external-link">BuildClusterTree</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/PlotClusterTree.html" class="external-link">PlotClusterTree</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClusterDE-PBMC_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>We perform the DE test between cluster 2 and 8. We subset the cluster
2 and 8 (<code>pbmc_sub</code>).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pbmc</span>, idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">original_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc_sub</span>, </span>
<span>                                ident.1 <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                                ident.2 <span class="op">=</span> <span class="fl">8</span>, </span>
<span>                                min.pct <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                                logfc.threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">#&gt; For a more efficient implementation of the Wilcoxon Rank Sum Test,</span></span>
<span><span class="co">#&gt; (default method for FindMarkers) please install the limma package</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; install.packages('BiocManager')</span></span>
<span><span class="co">#&gt; BiocManager::install('limma')</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; After installation of limma, Seurat will automatically use the more </span></span>
<span><span class="co">#&gt; efficient implementation (no further action necessary).</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="generate-synthetic-null-data">Generate synthetic null data<a class="anchor" aria-label="anchor" href="#generate-synthetic-null-data"></a>
</h2>
<p>We first generate the synthetic null data based on the target data
(real subset data, <code>pbmc_sub</code>, which contains the two
clusters you are interested in). You can increase the number of cores to
speed it up.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_sub</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">synthetic_null</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="../reference/constructNull.html">constructNull</a></span><span class="op">(</span><span class="va">count_mat</span>, nCores <span class="op">=</span> <span class="fl">2</span>, fastVersion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt; 107.129  35.044  74.955</span></span></code></pre></div>
<p>We perform the same pipeline as we did for target data. Please note
we need two clusters here, too. This is straightforward for clustering
method with predefined cluster numbers, e.g., K-means. For louvain
clustering used in Seurat you may manually adjust the resolution
parameter so that we still get two clusters.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">synthetic_null</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span><span class="op">)</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span><span class="op">)</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span><span class="op">)</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span><span class="op">)</span></span>
<span><span class="co">#&gt; PC_ 1 </span></span>
<span><span class="co">#&gt; Positive:  RPS19, RPS27, RPL41, AIF1, LST1, B2M, RPL10, COTL1, FTH1, RPL19 </span></span>
<span><span class="co">#&gt;     RPL8, RPL11, PFN1, HLA-C, NACA, YBX1, EEF1A1, FTL, FCER1G, HLA-B </span></span>
<span><span class="co">#&gt;     RPL7A, TMSB10, IFITM3, RPL14, RPL15, FAU, CORO1A, RPS2, TMSB4X, RNASET2 </span></span>
<span><span class="co">#&gt; Negative:  VCAN, S100A8, SLC2A3, FOS, S100A9, S100A12, CD14, LYZ, ITGAM, CSF3R </span></span>
<span><span class="co">#&gt;     CD36, SELL, NCF1, MS4A6A, RGS2, DUSP6, GPX1, IRF2BP2, FPR1, FOSB </span></span>
<span><span class="co">#&gt;     NR4A2, THBS1, ALOX5AP, CCR1, RNASE6, RP11-1143G9.4, ZFP36L1, NRGN, CYP1B1, CLEC4E </span></span>
<span><span class="co">#&gt; PC_ 2 </span></span>
<span><span class="co">#&gt; Positive:  CDKN1C, RHOC, IFITM2, FCGR3A, KLF2, POU2F2, PAG1, HES4, MS4A7, SLC2A6 </span></span>
<span><span class="co">#&gt;     LRRC25, PECAM1, RRAS, FAM110A, CYTIP, LYST, SSH2, DRAP1, LYL1, CSF1R </span></span>
<span><span class="co">#&gt;     IFITM3, MTSS1, SPN, HN1, UNC119, SSBP4, SIGLEC10, SNX9, SOD1, BID </span></span>
<span><span class="co">#&gt; Negative:  TPT1, RPS13, GAPDH, LYZ, GPX1, RPS18, RPS3A, RPL18A, VIM, RPL32 </span></span>
<span><span class="co">#&gt;     RPL3, S100A10, RPS23, RPS6, RPL6, S100A9, RPS8, GRN, EEF1A1, RPLP0 </span></span>
<span><span class="co">#&gt;     CD14, LGALS3, RPL34, RPL30, AP1S2, FCN1, FOS, RPL23A, RPS12, RPS4X </span></span>
<span><span class="co">#&gt; PC_ 3 </span></span>
<span><span class="co">#&gt; Positive:  HMGB2, RGS2, RPS26, IRS2, TNFSF10, RNF144B, S100A12, CBX4, RPL34, RPL14 </span></span>
<span><span class="co">#&gt;     S100A9, SOD2, S100A8, ACSL1, SELL, SOX4, ALOX5AP, EXOC1, EGR1, RPL30 </span></span>
<span><span class="co">#&gt;     RPS3, RPS12, VCAN, CD48, FAM53C, CLEC2B, RPL12, GNG2, SARAF, SERTAD1 </span></span>
<span><span class="co">#&gt; Negative:  HLA-DRA, CPVL, HLA-DPA1, HLA-DPB1, MAP3K8, HLA-DRB1, MARCKS, CD74, XYLT1, EZR </span></span>
<span><span class="co">#&gt;     CST3, HLA-DQB1, HMOX1, CD300E, NAAA, BHLHE40, CD63, DUSP6, CDKN1A, RGCC </span></span>
<span><span class="co">#&gt;     MAFB, APOBEC3A, HLA-DMA, DOK2, EMP3, RUNX3, ID2, ZMIZ1, LGALS3, CD93 </span></span>
<span><span class="co">#&gt; PC_ 4 </span></span>
<span><span class="co">#&gt; Positive:  PLBD1, PLAC8, S100A8, S100A9, CES1, SLA, S100A12, PARP14, SELL, PPP3R1 </span></span>
<span><span class="co">#&gt;     SOD2, MAFB, FCN1, IRF7, BCL2A1, LY86, SPTLC2, OSM, VCAN, LYN </span></span>
<span><span class="co">#&gt;     TUBA4A, AP1S2, LYZ, MBD2, OAS1, BLOC1S1, FPR1, APOBEC3A, RP11-1143G9.4, CD14 </span></span>
<span><span class="co">#&gt; Negative:  RPL34, CDC123, RPS23, RPL12, HLA-DMB, RPL32, ANKRD36C, RPL14, RPL30, LPAR6 </span></span>
<span><span class="co">#&gt;     EEF1B2, AHR, RPS4X, UBE2D4, TNNI2, HLA-DPB1, HIST1H1C, RSPRY1, HLA-DPA1, SNTB1 </span></span>
<span><span class="co">#&gt;     HMGN1, RPS3A, MGAT4B, RPS12, FAM78A, RPS18, ICOSLG, CKS1B, B9D2, DTYMK </span></span>
<span><span class="co">#&gt; PC_ 5 </span></span>
<span><span class="co">#&gt; Positive:  S100A11, ACTB, GALNS, PLPPR2, FTH1, FTL, B3GALT6, TFAP4, ARPC3, DUSP10 </span></span>
<span><span class="co">#&gt;     LCK, RALGDS, ZNF281, TYROBP, COTL1, ACTG1, GAPDH, S100A12, PLAUR, CXCL8 </span></span>
<span><span class="co">#&gt;     BCL2A1, S100A4, NSFL1C, LILRB2, PLBD1, IL1RAP, CES1, PDE4B, FCER1G, APH1B </span></span>
<span><span class="co">#&gt; Negative:  IFI44, STOM, ZC3HAV1, ARL4C, ATM, IFI44L, APOL6, MYLIP, MX1, AKNA </span></span>
<span><span class="co">#&gt;     TNFAIP3, PARP14, STAT1, C5orf24, CCPG1, ZCCHC2, SNRNP200, MZT2B, DDX6, MGAT1 </span></span>
<span><span class="co">#&gt;     CMPK2, PRAM1, SHISA5, UTRN, TMEM70, FBXO7, DNAJC8, MIS18BP1, B2M, MECP2</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; Computing SNN</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span>, resolution <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 453</span></span>
<span><span class="co">#&gt; Number of edges: 17874</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.7632</span></span>
<span><span class="co">#&gt; Number of communities: 2</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; 18:17:52 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; 18:17:52 Read 453 rows and found 10 numeric columns</span></span>
<span><span class="co">#&gt; 18:17:52 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; 18:17:52 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 18:17:52 Writing NN index file to temp file /tmp/RtmptRhvxT/file1b962d300176</span></span>
<span><span class="co">#&gt; 18:17:52 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 18:17:52 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 18:17:53 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 18:17:54 Initializing from normalized Laplacian + noise (using irlba)</span></span>
<span><span class="co">#&gt; 18:17:54 Commencing optimization for 500 epochs, with 18194 positive edges</span></span>
<span><span class="co">#&gt; 18:17:55 Optimization finished</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClusterDE-PBMC_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>We perform the DE test on synthetic null data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">null_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc_null</span>, </span>
<span>                            ident.1 <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                            ident.2 <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                            min.pct <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                            logfc.threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>We extract the p-values from both original data and synthetic null
data, then use <code>ClusterDE</code> to “compare” them.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">original_pval</span> <span class="op">&lt;-</span> <span class="va">original_markers</span><span class="op">$</span><span class="va">p_val</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">original_pval</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">original_markers</span><span class="op">)</span></span>
<span></span>
<span><span class="va">null_pval</span> <span class="op">&lt;-</span> <span class="va">null_markers</span><span class="op">$</span><span class="va">p_val</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">null_pval</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">null_markers</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="../reference/callDE.html">callDE</a></span><span class="op">(</span><span class="va">original_pval</span>, <span class="va">null_pval</span>, nlogTrans <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>You can check the summary table. The genes are ranked by “contrast
score (cs)”.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">summaryTable</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   Gene   target    null    cs      q</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> CDKN1C   67.5 12.9     54.6 0.047<span style="text-decoration: underline;">6</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> CD79B    53.1  0.080<span style="text-decoration: underline;">2</span>  53.1 0.047<span style="text-decoration: underline;">6</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> FCGR3A   79.0 27.1     51.9 0.047<span style="text-decoration: underline;">6</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> CKB      44.6  0.916   43.6 0.047<span style="text-decoration: underline;">6</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> RHOC     65.8 23.0     42.8 0.047<span style="text-decoration: underline;">6</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> PLD4     36.2  0.942   35.2 0.047<span style="text-decoration: underline;">6</span></span></span></code></pre></div>
<p>To compare the result from the naive Seurat pipeline and ClusterDE,
we first visualize the top 6 DE genes from Seurat. Genes <em>LST1</em>
and <em>RPS19</em> are both highly expressed in two clusters. In
addition, <em>RPS19</em> is reported as a stable housekeeping genes in
several studies. Note that it does not mean the expression levels of
<em>LST1</em> and <em>RPS19</em> are the same between the two cell
types. It means that they are not good cell type markers.
Philosophically speaking, it means that <strong>conditional on the two
clusters are obtained by clustering algorithm, LST1 and RPS19 are less
likely to be the cell type markers between the two cell
types</strong>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">[</span> , <span class="va">pbmc</span><span class="op">$</span><span class="va">seurat_clusters</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">8</span><span class="op">)</span><span class="op">]</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">original_markers</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClusterDE-PBMC_files/figure-html/unnamed-chunk-11-1.png" width="960">
In contrast, the genes from ClusterDE do not have <em>LST1</em> and
<em>RPS19</em> anymore.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">[</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">seurat_clusters</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">8</span><span class="op">)</span><span class="op">]</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">summaryTable</span><span class="op">$</span><span class="va">Gene</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClusterDE-PBMC_files/figure-html/unnamed-chunk-12-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Random number generation:</span></span>
<span><span class="co">#&gt;  RNG:     L'Ecuyer-CMRG </span></span>
<span><span class="co">#&gt;  Normal:  Inversion </span></span>
<span><span class="co">#&gt;  Sample:  Rejection </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] Matrix_1.6-1.1              patchwork_1.1.3            </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.4.3               SingleCellExperiment_1.22.0</span></span>
<span><span class="co">#&gt;  [5] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">#&gt;  [7] GenomicRanges_1.52.1        GenomeInfoDb_1.36.4        </span></span>
<span><span class="co">#&gt;  [9] IRanges_2.34.1              S4Vectors_0.38.2           </span></span>
<span><span class="co">#&gt; [11] BiocGenerics_0.46.0         MatrixGenerics_1.12.3      </span></span>
<span><span class="co">#&gt; [13] matrixStats_1.0.0           SeuratObject_4.1.4         </span></span>
<span><span class="co">#&gt; [15] Seurat_4.4.0                ClusterDE_0.99.1           </span></span>
<span><span class="co">#&gt; [17] BiocStyle_2.28.1           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3      jsonlite_1.8.7          magrittr_2.0.3         </span></span>
<span><span class="co">#&gt;   [4] spatstat.utils_3.0-3    farver_2.1.1            rmarkdown_2.25         </span></span>
<span><span class="co">#&gt;   [7] zlibbioc_1.46.0         fs_1.6.3                ragg_1.2.6             </span></span>
<span><span class="co">#&gt;  [10] vctrs_0.6.3             ROCR_1.0-11             memoise_2.0.1          </span></span>
<span><span class="co">#&gt;  [13] spatstat.explore_3.2-3  RCurl_1.98-1.12         S4Arrays_1.0.6         </span></span>
<span><span class="co">#&gt;  [16] htmltools_0.5.6.1       sass_0.4.7              sctransform_0.4.0      </span></span>
<span><span class="co">#&gt;  [19] parallelly_1.36.0       KernSmooth_2.23-21      bslib_0.5.1            </span></span>
<span><span class="co">#&gt;  [22] htmlwidgets_1.6.2       desc_1.4.2              ica_1.0-3              </span></span>
<span><span class="co">#&gt;  [25] plyr_1.8.9              plotly_4.10.2           zoo_1.8-12             </span></span>
<span><span class="co">#&gt;  [28] cachem_1.0.8            igraph_1.5.1            mime_0.12              </span></span>
<span><span class="co">#&gt;  [31] lifecycle_1.0.3         pkgconfig_2.0.3         R6_2.5.1               </span></span>
<span><span class="co">#&gt;  [34] fastmap_1.1.1           GenomeInfoDbData_1.2.10 fitdistrplus_1.1-11    </span></span>
<span><span class="co">#&gt;  [37] future_1.33.0           shiny_1.7.5             digest_0.6.33          </span></span>
<span><span class="co">#&gt;  [40] colorspace_2.1-0        rprojroot_2.0.3         tensor_1.5             </span></span>
<span><span class="co">#&gt;  [43] irlba_2.3.5.1           kde1d_1.0.5             textshaping_0.3.7      </span></span>
<span><span class="co">#&gt;  [46] labeling_0.4.3          progressr_0.14.0        fansi_1.0.5            </span></span>
<span><span class="co">#&gt;  [49] spatstat.sparse_3.0-2   httr_1.4.7              polyclip_1.10-6        </span></span>
<span><span class="co">#&gt;  [52] abind_1.4-5             compiler_4.3.1          withr_2.5.1            </span></span>
<span><span class="co">#&gt;  [55] MASS_7.3-60             DelayedArray_0.26.7     tools_4.3.1            </span></span>
<span><span class="co">#&gt;  [58] lmtest_0.9-40           ape_5.7-1               httpuv_1.6.11          </span></span>
<span><span class="co">#&gt;  [61] future.apply_1.11.0     goftest_1.2-3           glue_1.6.2             </span></span>
<span><span class="co">#&gt;  [64] nlme_3.1-162            promises_1.2.1          grid_4.3.1             </span></span>
<span><span class="co">#&gt;  [67] Rtsne_0.16              cluster_2.1.4           reshape2_1.4.4         </span></span>
<span><span class="co">#&gt;  [70] generics_0.1.3          gtable_0.3.4            spatstat.data_3.0-1    </span></span>
<span><span class="co">#&gt;  [73] tidyr_1.3.0             data.table_1.14.8       XVector_0.40.0         </span></span>
<span><span class="co">#&gt;  [76] sp_2.1-0                utf8_1.2.3              spatstat.geom_3.2-5    </span></span>
<span><span class="co">#&gt;  [79] RcppAnnoy_0.0.21        ggrepel_0.9.3           RANN_2.6.1             </span></span>
<span><span class="co">#&gt;  [82] pillar_1.9.0            stringr_1.5.0           later_1.3.1            </span></span>
<span><span class="co">#&gt;  [85] splines_4.3.1           dplyr_1.1.3             lattice_0.21-8         </span></span>
<span><span class="co">#&gt;  [88] survival_3.5-5          deldir_1.0-9            tidyselect_1.2.0       </span></span>
<span><span class="co">#&gt;  [91] miniUI_0.1.1.1          pbapply_1.7-2           knitr_1.44             </span></span>
<span><span class="co">#&gt;  [94] gridExtra_2.3           bookdown_0.35           scattermore_1.2        </span></span>
<span><span class="co">#&gt;  [97] xfun_0.40               stringi_1.7.12          lazyeval_0.2.2         </span></span>
<span><span class="co">#&gt; [100] yaml_2.3.7              evaluate_0.22           codetools_0.2-19       </span></span>
<span><span class="co">#&gt; [103] tibble_3.2.1            BiocManager_1.30.22     cli_3.6.1              </span></span>
<span><span class="co">#&gt; [106] uwot_0.1.16             xtable_1.8-4            reticulate_1.32.0      </span></span>
<span><span class="co">#&gt; [109] systemfonts_1.0.5       munsell_0.5.0           jquerylib_0.1.4        </span></span>
<span><span class="co">#&gt; [112] Rcpp_1.0.11             globals_0.16.2          spatstat.random_3.1-6  </span></span>
<span><span class="co">#&gt; [115] png_0.1-8               rngWELL_0.10-9          parallel_4.3.1         </span></span>
<span><span class="co">#&gt; [118] ellipsis_0.3.2          randtoolbox_2.0.4       assertthat_0.2.1       </span></span>
<span><span class="co">#&gt; [121] pkgdown_2.0.7           mvnfast_0.2.8           bitops_1.0-7           </span></span>
<span><span class="co">#&gt; [124] listenv_0.9.0           viridisLite_0.4.2       scales_1.2.1           </span></span>
<span><span class="co">#&gt; [127] ggridges_0.5.4          crayon_1.5.2            leiden_0.4.3           </span></span>
<span><span class="co">#&gt; [130] purrr_1.0.2             rlang_1.1.1             rvinecopulib_0.6.3.1.1 </span></span>
<span><span class="co">#&gt; [133] cowplot_1.1.1</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Dongyuan Song, Kexin Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
