<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Perform ClusterDE on a PBMC dataset • ClusterDE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform ClusterDE on a PBMC dataset">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ClusterDE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/ClusterDE-cellline.html">Perform ClusterDE on a cell line dataset</a></li>
    <li><a class="dropdown-item" href="../articles/ClusterDE-onedomain.html">Perform ClusterDE on a single-domain spatial dataset</a></li>
    <li><a class="dropdown-item" href="../articles/ClusterDE-PBMC.html">Perform ClusterDE on a PBMC dataset</a></li>
    <li><a class="dropdown-item" href="../articles/ClusterDE-twodomains.html">Perform ClusterDE on a two-domain spatial dataset</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/SONGDONGYUAN1994/ClusterDE/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Perform ClusterDE on a PBMC dataset</h1>
                        <h4 data-toc-skip class="author">Dongyuan
Song</h4>
            <address class="author_afil">
      Bioinformatics IDP, University of California, Los
Angeles<br><a class="author_email" href="mailto:#"></a><a href="mailto:dongyuansong@ucla.edu" class="email">dongyuansong@ucla.edu</a>
      </address>
                  
            <h4 data-toc-skip class="date">31 July 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/SONGDONGYUAN1994/ClusterDE/blob/main/vignettes/ClusterDE-PBMC.Rmd" class="external-link"><code>vignettes/ClusterDE-PBMC.Rmd</code></a></small>
      <div class="d-none name"><code>ClusterDE-PBMC.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#tools::R_user_dir("ClusterDE", which="cache")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SONGDONGYUAN1994/ClusterDE" class="external-link">ClusterDE</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="co">#library(SeuratData)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="download-data">Download data<a class="anchor" aria-label="anchor" href="#download-data"></a>
</h2>
<p>The PBMC datasets are originally from <code>SeuratData</code>. We use
one of them (10x Chromium (v3) from PBMC1 replicate). We filtered out
some lowly epxressed genes to save computational time here.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># InstallData("pbmcsca")</span></span>
<span><span class="co"># data("pbmcsca")</span></span>
<span><span class="co"># pbmc &lt;- pbmcsca[, pbmcsca@meta.data$Method=="10x Chromium (v3)" &amp; pbmcsca@meta.data$Experiment == "pbmc1"]</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># pbmc &lt;- pbmc[Matrix::rowSums(pbmc@assays$RNA@counts) &gt; 100, ]</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"https://figshare.com/ndownloader/files/41486283"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-the-regular-seurat-pipeline">Run the regular Seurat pipeline<a class="anchor" aria-label="anchor" href="#run-the-regular-seurat-pipeline"></a>
</h2>
<p>We perform the default Seurat clustering. Note that in real data
analysis, the cell type label is usually unknown.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">RNGkind</a></span><span class="op">(</span><span class="st">"L'Ecuyer-CMRG"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="co">## Avoid errors</span></span>
<span><span class="co">#&gt; Validating object structure</span></span>
<span><span class="co">#&gt; Updating object slots</span></span>
<span><span class="co">#&gt; Ensuring keys are in the proper structure</span></span>
<span><span class="co">#&gt; Warning: Assay RNA changing from Assay to Assay</span></span>
<span><span class="co">#&gt; Ensuring keys are in the proper structure</span></span>
<span><span class="co">#&gt; Ensuring feature names don't have underscores or pipes</span></span>
<span><span class="co">#&gt; Updating slots in RNA</span></span>
<span><span class="co">#&gt; Validating object structure for Assay 'RNA'</span></span>
<span><span class="co">#&gt; Object representation is consistent with the most current Seurat version</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="co">#&gt; PC_ 1 </span></span>
<span><span class="co">#&gt; Positive:  IL32, CCL5, TRBC2, TRAC, CD69, CST7, RORA, CTSW, SPOCK2, ITM2A </span></span>
<span><span class="co">#&gt;     GZMM, CD247, TRBC1, C12orf75, IL7R, CD8A, CD2, LDHB, GZMA, CD7 </span></span>
<span><span class="co">#&gt;     NKG7, CD6, GZMH, CD8B, BCL11B, PRF1, LYAR, LTB, FGFBP2, TCF7 </span></span>
<span><span class="co">#&gt; Negative:  LYZ, FCN1, CLEC7A, CPVL, SERPINA1, SPI1, S100A9, AIF1, NAMPT, CSTA </span></span>
<span><span class="co">#&gt;     CTSS, MAFB, MPEG1, NCF2, VCAN, FGL2, S100A8, TYMP, CST3, LST1 </span></span>
<span><span class="co">#&gt;     CYBB, CFD, FCER1G, SLC11A1, TGFBI, GRN, CD14, PSAP, SLC7A7, MS4A6A </span></span>
<span><span class="co">#&gt; PC_ 2 </span></span>
<span><span class="co">#&gt; Positive:  RPL10, EEF1A1, TMSB10, RPS2, RPS12, RPL13, RPS18, RPS23, RPLP1, TPT1 </span></span>
<span><span class="co">#&gt;     RPS8, IL32, S100A4, PFN1, RPLP0, NKG7, ARL4C, HSPA8, CST7, ZFP36L2 </span></span>
<span><span class="co">#&gt;     ANXA1, CTSW, S100A6, LDHA, CORO1A, CD247, GZMA, CALR, S100A10, GZMM </span></span>
<span><span class="co">#&gt; Negative:  NRGN, PF4, SDPR, HIST1H2AC, MAP3K7CL, PPBP, GNG11, GPX1, TUBB1, SPARC </span></span>
<span><span class="co">#&gt;     CLU, PGRMC1, FTH1, RGS18, MARCH2, TREML1, HIST1H3H, AP003068.23, NCOA4, ACRBP </span></span>
<span><span class="co">#&gt;     TAGLN2, PRKAR2B, CD9, CA2, CMTM5, CTTN, MTURN, TMSB4X, HIST1H2BJ, TSC22D1 </span></span>
<span><span class="co">#&gt; PC_ 3 </span></span>
<span><span class="co">#&gt; Positive:  CD79A, HLA-DQA1, MS4A1, LINC00926, IGHM, BANK1, IGHD, TNFRSF13C, HLA-DQB1, CD74 </span></span>
<span><span class="co">#&gt;     IGKC, HLA-DRA, BLK, CD83, CD37, CD22, ADAM28, JUND, NFKBID, HLA-DRB1 </span></span>
<span><span class="co">#&gt;     P2RX5, CD79B, VPREB3, IGLC2, FCER2, RPS8, LTB, RPS23, TCOF1, GNG7 </span></span>
<span><span class="co">#&gt; Negative:  CCL5, TMSB4X, SRGN, NKG7, ACTB, CST7, GZMH, FGFBP2, CTSW, PRF1 </span></span>
<span><span class="co">#&gt;     GZMA, GZMB, C12orf75, S100A4, ANXA1, KLRD1, NRGN, GNLY, GZMM, IL32 </span></span>
<span><span class="co">#&gt;     PF4, SDPR, PPBP, MYO1F, CD247, GAPDH, MAP3K7CL, HIST1H2AC, GNG11, TUBB1 </span></span>
<span><span class="co">#&gt; PC_ 4 </span></span>
<span><span class="co">#&gt; Positive:  FCGR3A, GZMB, FGFBP2, GZMH, NKG7, HLA-DPA1, PRF1, HLA-DPB1, CST7, GNLY </span></span>
<span><span class="co">#&gt;     KLRD1, HLA-DRB1, GZMA, CCL5, SPON2, ADGRG1, CTSW, ZEB2, PRSS23, IFITM2 </span></span>
<span><span class="co">#&gt;     CCL4, CD74, KLRF1, RHOC, MTSS1, CDKN1C, CD79B, CEP78, HLA-DQA1, CLIC3 </span></span>
<span><span class="co">#&gt; Negative:  IL7R, LEPROTL1, LTB, RCAN3, MAL, LEF1, TCF7, ZFP36L2, CAMK4, VIM </span></span>
<span><span class="co">#&gt;     LDHB, NOSIP, JUNB, SLC2A3, TRABD2A, RGCC, SATB1, TNFAIP3, TMEM123, SOCS3 </span></span>
<span><span class="co">#&gt;     AQP3, BCL11B, NELL2, TNFRSF25, CD28, PABPC1, DNAJB1, TRAT1, OXNAD1, TRAC </span></span>
<span><span class="co">#&gt; PC_ 5 </span></span>
<span><span class="co">#&gt; Positive:  CDKN1C, HES4, CSF1R, CKB, ZNF703, TCF7L2, CTSL, MS4A7, PAG1, FAM110A </span></span>
<span><span class="co">#&gt;     SIGLEC10, LRRC25, FCGR3A, LTB, RNASET2, CDH23, IL7R, RRAS, LINC01272, IFITM3 </span></span>
<span><span class="co">#&gt;     LST1, LILRB2, PILRA, RHOC, SLC2A6, PECAM1, CAMK1, TAGLN, IFI30, BID </span></span>
<span><span class="co">#&gt; Negative:  VCAN, S100A12, S100A8, CD14, CSF3R, ITGAM, CST7, GZMB, MT-CO1, GNLY </span></span>
<span><span class="co">#&gt;     KLRD1, PRF1, MS4A6A, GZMH, FGFBP2, CD93, EGR1, NKG7, S100A9, MT-CO3 </span></span>
<span><span class="co">#&gt;     IER3, THBS1, RNASE6, CLEC4E, MGST1, CTSW, SGK1, GZMA, RP11-1143G9.4, CH17-373J23.1</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; Computing SNN</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, resolution <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 3222</span></span>
<span><span class="co">#&gt; Number of edges: 108605</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.9363</span></span>
<span><span class="co">#&gt; Number of communities: 10</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="co">#&gt; 00:44:42 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; 00:44:42 Read 3222 rows and found 10 numeric columns</span></span>
<span><span class="co">#&gt; 00:44:42 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; 00:44:42 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 00:44:42 Writing NN index file to temp file /tmp/RtmprOTKJf/file83daa2b7baa</span></span>
<span><span class="co">#&gt; 00:44:42 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 00:44:43 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 00:44:44 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 00:44:45 Initializing from normalized Laplacian + noise (using RSpectra)</span></span>
<span><span class="co">#&gt; 00:44:45 Commencing optimization for 500 epochs, with 126446 positive edges</span></span>
<span><span class="co">#&gt; 00:44:48 Optimization finished</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Clustering result"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"CellType"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="ClusterDE-PBMC_files/figure-html/unnamed-chunk-3-1.png" width="1152"></p>
<p>In this vignette, we are interested in cluster 2 vs 8, which
approximately represent CD14+/CD16+ monocytes. Please note that
ClusterDE is designed for 1 vs 1 comparison. Therefore, users may (1)
choose the two interested clusters manually based on their knowledge or
(2) use the two locally closest clusters from computation (e.g.,
<code>BuildClusterTree</code> in <code>Seurat</code>).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/BuildClusterTree.html" class="external-link">BuildClusterTree</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/PlotClusterTree.html" class="external-link">PlotClusterTree</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClusterDE-PBMC_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>We perform the DE test between cluster 2 and 8. We subset the cluster
2 and 8 (<code>pbmc_sub</code>).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pbmc</span>, idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">original_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc_sub</span>, </span>
<span>                                ident.1 <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                                ident.2 <span class="op">=</span> <span class="fl">8</span>, </span>
<span>                                min.pct <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                                logfc.threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">#&gt; For a (much!) faster implementation of the Wilcoxon Rank Sum Test,</span></span>
<span><span class="co">#&gt; (default method for FindMarkers) please install the presto package</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; install.packages('devtools')</span></span>
<span><span class="co">#&gt; devtools::install_github('immunogenomics/presto')</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; After installation of presto, Seurat will automatically use the more </span></span>
<span><span class="co">#&gt; efficient implementation (no further action necessary).</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span></code></pre></div>
<p><strong>Note</strong>: since we need the same number of genes being
tested in real data and null data, we manually remove the default
filtering in Seurat, which will exclude some genes from being test. We
believe that this fitering should be performed after you test every
gene.</p>
</div>
<div class="section level2">
<h2 id="generate-synthetic-null-data">Generate synthetic null data<a class="anchor" aria-label="anchor" href="#generate-synthetic-null-data"></a>
</h2>
<p>We first generate the synthetic null data based on the target data
(real subset data, <code>pbmc_sub</code>, which contains the two
clusters you are interested in). You can increase the number of cores
(CPUs) to speed it up.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_sub</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">synthetic_null</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="../reference/constructNull.html">constructNull</a></span><span class="op">(</span><span class="va">count_mat</span>, nCores <span class="op">=</span> <span class="fl">2</span>, fastVersion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 130 genes have no more than 2 non-zero values; ignore fitting and return all 0s.</span></span>
<span><span class="co">#&gt; 64.5% of genes are used in correlation modelling.</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;  84.205  55.444  39.975</span></span></code></pre></div>
<p>We perform the same pipeline as we did for target data. Please note
we need two clusters here, too. This is straightforward for clustering
method with predefined cluster numbers, e.g., K-means. For louvain
clustering used in Seurat you may manually adjust the resolution
parameter so that we still get two clusters.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">synthetic_null</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span><span class="op">)</span></span>
<span><span class="co">#&gt; Normalizing layer: counts</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding variable features for layer counts</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span><span class="op">)</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span><span class="op">)</span></span>
<span><span class="co">#&gt; PC_ 1 </span></span>
<span><span class="co">#&gt; Positive:  RPS19, PFN1, AIF1, RPS27, LST1, RPL8, YBX1, COTL1, RPL41, RNASET2 </span></span>
<span><span class="co">#&gt;     FTL, IFITM3, FCGR3A, RPL10, IFITM2, FTH1, NACA, FCER1G, TMSB4X, SOD1 </span></span>
<span><span class="co">#&gt;     RPS29, B2M, GNAI2, HLA-B, RHOC, HLA-C, TMSB10, FAU, RPL19, MS4A7 </span></span>
<span><span class="co">#&gt; Negative:  VCAN, FOS, S100A8, LYZ, S100A9, SLC2A3, CD14, CSF3R, S100A12, IRF2BP2 </span></span>
<span><span class="co">#&gt;     GPX1, MS4A6A, ITGAM, FOSB, DUSP6, CD36, CYP1B1, PPIF, SELL, RP11-1143G9.4 </span></span>
<span><span class="co">#&gt;     ZFP36L1, NFKBIA, RGS2, CD93, SGK1, RNASE6, CD99, STAB1, NCF1, CEBPD </span></span>
<span><span class="co">#&gt; PC_ 2 </span></span>
<span><span class="co">#&gt; Positive:  CDKN1C, MALAT1, FAM110A, SLC2A6, FCGR3A, MTSS1, RRAS, TCF7L2, KLF2, PECAM1 </span></span>
<span><span class="co">#&gt;     PAG1, LYST, RHOC, SOD1, ABI3, CYFIP2, CKB, SPN, TPTEP1, ADA </span></span>
<span><span class="co">#&gt;     HES4, CDH23, LTB, SLC44A2, PHTF2, PIK3CG, CAT, ITGAX, SNX9, CD79B </span></span>
<span><span class="co">#&gt; Negative:  TPT1, RPS13, GAPDH, RPS3A, RPS18, RPLP0, HLA-DRA, RPL18A, EEF1A1, RPS2 </span></span>
<span><span class="co">#&gt;     GPX1, HLA-DRB1, RPL37A, CPVL, RPS8, RPL28, RPS6, CD74, RPL23A, RPL3 </span></span>
<span><span class="co">#&gt;     RPS14, RPS4X, RPL26, RPL11, RPS12, FCN1, RPL37, RPL30, LYZ, RPL10 </span></span>
<span><span class="co">#&gt; PC_ 3 </span></span>
<span><span class="co">#&gt; Positive:  HLA-DRA, MARCKSL1, HLA-DPB1, HLA-DPA1, APOBEC3A, HLA-DRB1, EMP3, CD74, MARCKS, CPVL </span></span>
<span><span class="co">#&gt;     CD300E, DUSP6, HLA-DMB, DOK2, CST3, HLA-DQB1, NAAA, CEBPB, RUNX3, SRC </span></span>
<span><span class="co">#&gt;     RGCC, MAFB, TIMP1, MAP3K8, TLE3, CDKN1A, UTRN, ZMIZ1, HMOX1, IFI30 </span></span>
<span><span class="co">#&gt; Negative:  HMGB2, FYN, RPS3, IRS2, S100A8, ZFP36L2, RPS15A, S100A9, ADD3, ALOX5AP </span></span>
<span><span class="co">#&gt;     S100A12, SELL, UBE2D1, FAM65B, RGS2, RPS4X, PYGL, SORL1, PLBD1, CREBRF </span></span>
<span><span class="co">#&gt;     SAMSN1, RPL30, YPEL3, RPS6, RPL3, SOD2, RNF144B, EMB, RPS12, CNOT6L </span></span>
<span><span class="co">#&gt; PC_ 4 </span></span>
<span><span class="co">#&gt; Positive:  MALAT1, ARL4C, KCNAB2, LUC7L3, NKG7, SFT2D2, LINC00657, C5orf56, FAM133B, ZNF292 </span></span>
<span><span class="co">#&gt;     SORL1, LPP, PRRC2B, ZNF652, KLF9, BTG1, TMC6, YTHDC1, DDX6, RPS6KA3 </span></span>
<span><span class="co">#&gt;     DDHD1, SIGIRR, TTC14, GADD45B, PHKB, KDM7A, HBEGF, DENND1B, LBR, TOB1 </span></span>
<span><span class="co">#&gt; Negative:  S100A4, FTL, S100A11, CXCL8, FTH1, SERPINA1, S100A8, NCF1, ACTG1, DUSP6 </span></span>
<span><span class="co">#&gt;     FAM26F, ACTB, RGS2, FAU, FOS, COTL1, S100A9, S100A6, DIS3L2, SNAP23 </span></span>
<span><span class="co">#&gt;     SAT1, BCL2A1, CFD, CTB-61M7.2, PRELID1, S100A12, PLAUR, TRAC, AIF1, RHOB </span></span>
<span><span class="co">#&gt; PC_ 5 </span></span>
<span><span class="co">#&gt; Positive:  TUBA1A, TNFAIP3, ZCCHC8, LINC00657, APOBEC3A, DDX6, CIRBP, PLAUR, VASP, SNRK </span></span>
<span><span class="co">#&gt;     KMT2A, S1PR3, INSIG1, BHLHE40, EHD1, B2M, LILRA1, ARHGAP15, RUNX3, PDE4B </span></span>
<span><span class="co">#&gt;     EML4, DSTN, KIAA0430, FAM107B, PTGS2, WDFY1, PSMD1, TMX4, RPL30, CTNNB1 </span></span>
<span><span class="co">#&gt; Negative:  FAM102B, CEP63, ANXA1, CACUL1, TIMP1, UTP14A, TYW5, KAT8, SNHG9, C6orf226 </span></span>
<span><span class="co">#&gt;     OAS1, JUN, IRF8, AATF, TMEM55B, NAGK, YIPF4, TAGLN2, YDJC, SERTAD3 </span></span>
<span><span class="co">#&gt;     ADAM28, POMT1, LRP10, LMNA, SNX10, LPAR6, RAB37, HSPBP1, LPP, TSPAN17</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; Computing SNN</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span>, resolution <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 453</span></span>
<span><span class="co">#&gt; Number of edges: 18383</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.7450</span></span>
<span><span class="co">#&gt; Number of communities: 2</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="va">pbmc_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; 00:45:55 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; 00:45:55 Read 453 rows and found 10 numeric columns</span></span>
<span><span class="co">#&gt; 00:45:55 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; 00:45:55 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 00:45:55 Writing NN index file to temp file /tmp/RtmprOTKJf/file83da7088961</span></span>
<span><span class="co">#&gt; 00:45:55 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 00:45:55 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 00:45:56 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 00:45:57 Initializing from normalized Laplacian + noise (using RSpectra)</span></span>
<span><span class="co">#&gt; 00:45:57 Commencing optimization for 500 epochs, with 17726 positive edges</span></span>
<span><span class="co">#&gt; 00:45:58 Optimization finished</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_null</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClusterDE-PBMC_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>We perform the DE test on synthetic null data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">null_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html" class="external-link">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc_null</span>, </span>
<span>                            ident.1 <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                            ident.2 <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                            min.pct <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                            logfc.threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>We extract the p-values from both original data and synthetic null
data, then use <code>ClusterDE</code> to “compare” them.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">original_pval</span> <span class="op">&lt;-</span> <span class="va">original_markers</span><span class="op">$</span><span class="va">p_val</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">original_pval</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">original_markers</span><span class="op">)</span></span>
<span></span>
<span><span class="va">null_pval</span> <span class="op">&lt;-</span> <span class="va">null_markers</span><span class="op">$</span><span class="va">p_val</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">null_pval</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">null_markers</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">ClusterDE</span><span class="fu">::</span><span class="fu"><a href="../reference/callDE.html">callDE</a></span><span class="op">(</span><span class="va">original_pval</span>, <span class="va">null_pval</span>, nlogTrans <span class="op">=</span> <span class="cn">TRUE</span>, FDR <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<p>You can check the summary table. The genes are ranked by “contrast
score (cs)”.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">summaryTable</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   Gene   target  null    cs      q</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> FCGR3A   79.0  45.1  33.9 0.012<span style="text-decoration: underline;">3</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> CKB      44.6  13.1  31.5 0.012<span style="text-decoration: underline;">3</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> CDKN1C   67.5  37.2  30.3 0.012<span style="text-decoration: underline;">3</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> CD79B    53.1  25.5  27.7 0.012<span style="text-decoration: underline;">3</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> RHOC     65.8  39.5  26.2 0.012<span style="text-decoration: underline;">3</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> PLD4     36.2  13.6  22.6 0.012<span style="text-decoration: underline;">3</span></span></span></code></pre></div>
<p>To compare the result from the naive Seurat pipeline and ClusterDE,
we first visualize the top 6 DE genes from Seurat. Genes <em>LST1</em>
and <em>RPS19</em> are both highly expressed in two clusters. In
addition, <em>RPS19</em> is reported as a stable housekeeping genes in
several studies. Note that it does not mean the expression levels of
<em>LST1</em> and <em>RPS19</em> are the same between the two cell
types. It means that they are not good cell type markers.
Philosophically speaking, it means that <strong>conditional on the two
clusters are obtained by clustering algorithm, LST1 and RPS19 are less
likely to be the cell type markers between the two cell
types</strong>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">[</span> , <span class="va">pbmc</span><span class="op">$</span><span class="va">seurat_clusters</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">8</span><span class="op">)</span><span class="op">]</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">original_markers</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClusterDE-PBMC_files/figure-html/unnamed-chunk-11-1.png" width="960">
In contrast, the genes from ClusterDE do not have <em>LST1</em> and
<em>RPS19</em> anymore.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">[</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">seurat_clusters</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">8</span><span class="op">)</span><span class="op">]</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">summaryTable</span><span class="op">$</span><span class="va">Gene</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="ClusterDE-PBMC_files/figure-html/unnamed-chunk-12-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Random number generation:</span></span>
<span><span class="co">#&gt;  RNG:     L'Ecuyer-CMRG </span></span>
<span><span class="co">#&gt;  Normal:  Inversion </span></span>
<span><span class="co">#&gt;  Sample:  Rejection </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] Matrix_1.7-0                patchwork_1.2.0            </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.5.1               SingleCellExperiment_1.26.0</span></span>
<span><span class="co">#&gt;  [5] SummarizedExperiment_1.34.0 Biobase_2.64.0             </span></span>
<span><span class="co">#&gt;  [7] GenomicRanges_1.56.1        GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">#&gt;  [9] IRanges_2.38.1              S4Vectors_0.42.1           </span></span>
<span><span class="co">#&gt; [11] BiocGenerics_0.50.0         MatrixGenerics_1.16.0      </span></span>
<span><span class="co">#&gt; [13] matrixStats_1.3.0           Seurat_5.1.0               </span></span>
<span><span class="co">#&gt; [15] SeuratObject_5.0.2          sp_2.1-4                   </span></span>
<span><span class="co">#&gt; [17] ClusterDE_0.99.3            BiocStyle_2.32.1           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RcppAnnoy_0.0.22        splines_4.4.1           later_1.3.2            </span></span>
<span><span class="co">#&gt;   [4] tibble_3.2.1            polyclip_1.10-7         fastDummies_1.7.3      </span></span>
<span><span class="co">#&gt;   [7] lifecycle_1.0.4         globals_0.16.3          lattice_0.22-6         </span></span>
<span><span class="co">#&gt;  [10] MASS_7.3-60.2           magrittr_2.0.3          limma_3.60.4           </span></span>
<span><span class="co">#&gt;  [13] plotly_4.10.4           sass_0.4.9              rmarkdown_2.27         </span></span>
<span><span class="co">#&gt;  [16] jquerylib_0.1.4         yaml_2.3.10             httpuv_1.6.15          </span></span>
<span><span class="co">#&gt;  [19] sctransform_0.4.1       spam_2.10-0             spatstat.sparse_3.1-0  </span></span>
<span><span class="co">#&gt;  [22] reticulate_1.38.0       cowplot_1.1.3           pbapply_1.7-2          </span></span>
<span><span class="co">#&gt;  [25] RColorBrewer_1.1-3      abind_1.4-5             zlibbioc_1.50.0        </span></span>
<span><span class="co">#&gt;  [28] Rtsne_0.17              purrr_1.0.2             kde1d_1.0.7            </span></span>
<span><span class="co">#&gt;  [31] GenomeInfoDbData_1.2.12 ggrepel_0.9.5           irlba_2.3.5.1          </span></span>
<span><span class="co">#&gt;  [34] listenv_0.9.1           spatstat.utils_3.0-5    goftest_1.2-3          </span></span>
<span><span class="co">#&gt;  [37] RSpectra_0.16-2         spatstat.random_3.3-1   fitdistrplus_1.2-1     </span></span>
<span><span class="co">#&gt;  [40] parallelly_1.38.0       pkgdown_2.1.0           leiden_0.4.3.1         </span></span>
<span><span class="co">#&gt;  [43] codetools_0.2-20        DelayedArray_0.30.1     tidyselect_1.2.1       </span></span>
<span><span class="co">#&gt;  [46] UCSC.utils_1.0.0        farver_2.1.2            randtoolbox_2.0.4      </span></span>
<span><span class="co">#&gt;  [49] spatstat.explore_3.3-1  jsonlite_1.8.8          progressr_0.14.0       </span></span>
<span><span class="co">#&gt;  [52] ggridges_0.5.6          survival_3.6-4          systemfonts_1.1.0      </span></span>
<span><span class="co">#&gt;  [55] tools_4.4.1             ragg_1.3.2              ica_1.0-3              </span></span>
<span><span class="co">#&gt;  [58] Rcpp_1.0.13             glue_1.7.0              gridExtra_2.3          </span></span>
<span><span class="co">#&gt;  [61] SparseArray_1.4.8       xfun_0.46               mvnfast_0.2.8          </span></span>
<span><span class="co">#&gt;  [64] dplyr_1.1.4             withr_3.0.0             BiocManager_1.30.23    </span></span>
<span><span class="co">#&gt;  [67] fastmap_1.2.0           fansi_1.0.6             digest_0.6.36          </span></span>
<span><span class="co">#&gt;  [70] R6_2.5.1                mime_0.12               textshaping_0.4.0      </span></span>
<span><span class="co">#&gt;  [73] colorspace_2.1-1        scattermore_1.2         tensor_1.5             </span></span>
<span><span class="co">#&gt;  [76] spatstat.data_3.1-2     utf8_1.2.4              tidyr_1.3.1            </span></span>
<span><span class="co">#&gt;  [79] generics_0.1.3          data.table_1.15.4       httr_1.4.7             </span></span>
<span><span class="co">#&gt;  [82] htmlwidgets_1.6.4       S4Arrays_1.4.1          rngWELL_0.10-9         </span></span>
<span><span class="co">#&gt;  [85] uwot_0.2.2              pkgconfig_2.0.3         gtable_0.3.5           </span></span>
<span><span class="co">#&gt;  [88] lmtest_0.9-40           XVector_0.44.0          htmltools_0.5.8.1      </span></span>
<span><span class="co">#&gt;  [91] dotCall64_1.1-1         bookdown_0.40           scales_1.3.0           </span></span>
<span><span class="co">#&gt;  [94] png_0.1-8               spatstat.univar_3.0-0   knitr_1.48             </span></span>
<span><span class="co">#&gt;  [97] reshape2_1.4.4          nlme_3.1-164            cachem_1.1.0           </span></span>
<span><span class="co">#&gt; [100] zoo_1.8-12              stringr_1.5.1           KernSmooth_2.23-24     </span></span>
<span><span class="co">#&gt; [103] parallel_4.4.1          miniUI_0.1.1.1          desc_1.4.3             </span></span>
<span><span class="co">#&gt; [106] pillar_1.9.0            grid_4.4.1              vctrs_0.6.5            </span></span>
<span><span class="co">#&gt; [109] RANN_2.6.1              promises_1.3.0          xtable_1.8-4           </span></span>
<span><span class="co">#&gt; [112] cluster_2.1.6           evaluate_0.24.0         cli_3.6.3              </span></span>
<span><span class="co">#&gt; [115] compiler_4.4.1          rlang_1.1.4             crayon_1.5.3           </span></span>
<span><span class="co">#&gt; [118] future.apply_1.11.2     labeling_0.4.3          plyr_1.8.9             </span></span>
<span><span class="co">#&gt; [121] fs_1.6.4                stringi_1.8.4           viridisLite_0.4.2      </span></span>
<span><span class="co">#&gt; [124] deldir_2.0-4            assertthat_0.2.1        rvinecopulib_0.6.3.1.1 </span></span>
<span><span class="co">#&gt; [127] munsell_0.5.1           lazyeval_0.2.2          coop_0.6-3             </span></span>
<span><span class="co">#&gt; [130] spatstat.geom_3.3-2     RcppHNSW_0.6.0          future_1.34.0          </span></span>
<span><span class="co">#&gt; [133] statmod_1.5.0           shiny_1.9.0             highr_0.11             </span></span>
<span><span class="co">#&gt; [136] ROCR_1.0-11             igraph_2.0.3            bslib_0.8.0            </span></span>
<span><span class="co">#&gt; [139] ape_5.8</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dongyuan Song, Kexin Li, Siqi Chen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
